### ISSUE TITLE
New rule proposal: no-one-use-vars

### ISSUE THREAD

**Issue by ilyavolodin** [eslint/eslint#12374]
**Please describe what the rule should do:**
Prevent creation of variables that only used once in the code. In most cases, if the variable is only used once, it's easier to inline the value instead of polluting the scope.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Warns about a potential error (problem)
[x] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

**Examples of incorrect code for this rule**

```js
var foo = 22;
console.log(foo);
```

```js
function foo(bar) {
  const baz = bar;
  return baz++;
}
```

```js
function foo(bar) {
  const { baz } = bar;
  console.log(baz);
```

```js
var foo = function(bar) {
  return bar;
}

foo();
```

**Examples of correct code for this rule**
```js
function foo(bar) {
  return bar;
}
```

```js
function foo(bar) {
  console.log(bar.baz);
}
```

```js
var foo = require('foo');

foo();
```

```js
import foo from 'foo';

foo();
```

```js
function foo(bar) {
  const baz = bar.value;
  console.log(baz + baz ** 2);
```

**Why should this rule be included in ESLint (instead of a plugin)?**
It's a styling preference rule. Creating an extra variable for one-time use requires switching context when reading the code and pollutes scope. Exception might be when you have a very long operation on the right hand-side of the assignment (like maybe chaining multiple function calls). An option could be create to allow such cases. Also an option can be create to ignore variables that have function assigned to them.

**Are you willing to submit a pull request to implement this rule?**
Yes

**Comment by ljharb** [comment]
A common reason for doing this is to give you an easy/clear place to pause with the debugger, or to separate the stack trace location for "creating the value" from "using the value".

Additionally, intermediate variables can often add clarity even when entirely unnecessary.

**Comment by kaicataldo** [comment]
> Additionally, intermediate variables can often add clarity even when entirely unnecessary.

I know I personally oftentimes use one-time-use intermediate variables for this reason. I'm not convinced it's worth enforcing a practice for this, though - it really feels like something that should be evaluated on a case-by-case basis to me. That being said, if others find value in this rule, I'm not opposed to it.

**Comment by ilyavolodin** [comment]
> Additionally, intermediate variables can often add clarity even when entirely unnecessary.

Agree. I don't think this rule is for everyone. This is more of testing the waters to see how others feel about it. Problem is, I keep finding myself needing to comment on PRs because people create variables for no good reason.

**Comment by dimaqq** [comment]
@ilyavolodin how would you autofix `name` in the code below?

```js
const Comp = () => {
  const [ctr, set_ctr] = useState(0);
  const name = useSelector(state => state.some.name);
  return <div onClick={() => str_ctr(ctr+1)}>{ ctr?name:"too late" }</div>;
};
```
